sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/AvroUtils.java

```diff
@@ -20,119 +20,412 @@
+        Lists.newArrayListWithExpectedSize(beamSchema.getFieldCount());
+    GenericRecordBuilder builder = new GenericRecordBuilder(avroSchema);
+    for (int i = 0; i < beamSchema.getFieldCount(); ++i) {
+      Schema.Field field = beamSchema.getField(i);
+      builder.set(
+          field.getName(),
+          genericFromBeamField(
+              field.getType(), avroSchema.getField(field.getName()).schema(), row.getValue(i)));
+    }
+    return builder.build();
+  }
 
-      case LONG:
-        return Schema.FieldType.INT64;
+  /** Converts AVRO schema to Beam field. */
+  private static Schema.FieldType toFieldType(TypeWithNullability type) {
+    Schema.FieldType fieldType = null;
+    org.apache.avro.Schema avroSchema = type.type;
+
+    LogicalType logicalType = LogicalTypes.fromSchema(avroSchema);
+    if (logicalType != null) {
+      if (logicalType instanceof LogicalTypes.Decimal) {
+        fieldType = FieldType.DECIMAL;
+      } else if (logicalType instanceof LogicalTypes.TimestampMillis) {
+        // TODO: There is a desire to move Beam schema DATETIME to a micros representation. When
+        // this is done, this logical type needs to be changed.
+        fieldType = FieldType.DATETIME;
+      }
+    }
+
+    if (fieldType == null) {
+      switch (type.type.getType()) {
+        case RECORD:
+          fieldType = Schema.FieldType.row(toBeamSchema(avroSchema));
+          break;
+
+        case ENUM:
+          fieldType = Schema.FieldType.STRING;
+          break;
+
+        case ARRAY:
+          Schema.FieldType elementType =
+              toFieldType(new TypeWithNullability(avroSchema.getElementType()));
+          fieldType = Schema.FieldType.array(elementType);
+          break;
+
+        case MAP:
+          fieldType =
+              Schema.FieldType.map(
+                  Schema.FieldType.STRING,
+                  toFieldType(new TypeWithNullability(avroSchema.getValueType())));
+          break;
+
+        case FIXED:
+          fieldType = Schema.FieldType.BYTES;
+          break;
+
+        case STRING:
+          fieldType = Schema.FieldType.STRING;
+          break;
+
+        case BYTES:
+          fieldType = Schema.FieldType.BYTES;
+          break;
+
+        case INT:
+          fieldType = Schema.FieldType.INT32;
+          break;
+
+        case LONG:
+          fieldType = Schema.FieldType.INT64;
+          break;
+
+        case FLOAT:
+          fieldType = Schema.FieldType.FLOAT;
+          break;
+
+        case DOUBLE:
+          fieldType = Schema.FieldType.DOUBLE;
+          break;
+
+        case BOOLEAN:
+          fieldType = Schema.FieldType.BOOLEAN;
+          break;
+
+        case UNION:
+          throw new RuntimeException("Can't convert 'union' to FieldType");
+
+        case NULL:
+          throw new RuntimeException("Can't convert 'null' to FieldType");
+
+        default:
+          throw new AssertionError("Unexpected AVRO Schema.Type: " + avroSchema.getType());
```

Comment:
nit: probably same here, more descriptive message saying why this is unexpected would help. I understand that this is the "this should not happen case", but who knows...